// graphql-toolkit/dist/commonjs/loaders/schema/schema-from-export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const AggregateError = require("aggregate-error");
const fs_1 = require("fs");
const path_1 = require("path");
const isValidPath = require("is-valid-path");
const graphql_1 = require("graphql");
const schema_from_typedefs_1 = require("./schema-from-typedefs");
const isGlob = require("is-glob");
const glob_1 = require("glob");
const epoxy_1 = require("../../epoxy");
class SchemaFromExport {
    static getFiles(globOrValidPath) {
        return (isGlob(globOrValidPath)
            ? glob_1.sync(globOrValidPath, {
                absolute: true,
                cwd: process.cwd(),
            })
            : [globOrValidPath]).filter(file => !file.includes('node_modules') && !file.endsWith('.d.ts') && !file.endsWith('.spec.ts'));
    }
    async resolveExports(file) {
        const fullPath = path_1.isAbsolute(file) ? file : path_1.resolve(process.cwd(), file);
        if (isValidPath(file) && fs_1.existsSync(fullPath) && path_1.extname(file) !== '.json' && !schema_from_typedefs_1.isGraphQLFile(fullPath)) {
            try {
                const fileExport = await eval(`require('${fullPath}')`);
                return { ok: true, export: fileExport };
            }
            catch (e) {
                return { ok: false };
            }
        }
        return { ok: false };
    }
    async canHandle(globOrValidPath) {
        const files = SchemaFromExport.getFiles(globOrValidPath);
        for (let file of files) {
            const fullPath = path_1.isAbsolute(file) ? file : path_1.resolve(process.cwd(), file);
            if (isValidPath(file) && fs_1.existsSync(fullPath) && path_1.extname(file) !== '.json' && !schema_from_typedefs_1.isGraphQLFile(fullPath)) {
                const validRequire = await this.resolveExports(file);
                if (!validRequire.ok) {
                    return false;
                }
                const fileExports = validRequire.export;
                const schema = await (fileExports.default || fileExports.schema || fileExports);
                if (this.isSchemaObject(schema) || this.isSchemaAst(schema) || this.isSchemaText(schema) || this.isWrappedSchemaJson(schema) || this.isSchemaJson(schema)) {
                    return true;
                }
                return false;
            }
        }
        return false;
    }
    async handle(globOrValidPath, _options) {
        const files = SchemaFromExport.getFiles(globOrValidPath);
        const filtered = await files.filter(async (file) => await this.canHandle(file));
        const schemas = await Promise.all(filtered.map(async (file) => {
            const fullPath = path_1.isAbsolute(file) ? file : path_1.resolve(process.cwd(), file);
            if (fs_1.existsSync(fullPath)) {
                try {
                    const exports = await eval(`require('${fullPath}')`);
                    if (exports) {
                        let rawExport = exports.default || exports.schema || exports;
                        if (rawExport) {
                            let schema = await rawExport;
                            schema = await (schema.default || schema.schema || schema);
                            try {
                                return await this.resolveSchema(schema);
                            }
                            catch (e) {
                                throw new Error('Exported schema must be of type GraphQLSchema, text, AST, or introspection JSON.');
                            }
                        }
                        else {
                            throw new Error(`Invalid export from export file ${fullPath}: missing default export or 'schema' export!`);
                        }
                    }
                    else {
                        throw new Error(`Invalid export from export file ${fullPath}: empty export!`);
                    }
                }
                catch (e) {
                    console.warn(`Invalid export from export file ${fullPath}: missing default export or 'schema' export!`);
                    throw new AggregateError([new Error(`Unable to load schema from file "${file}" due to import error: ${e}`), e]);
                }
            }
            else {
                throw new Error(`Unable to locate introspection from export file: ${fullPath}`);
            }
        }));
        if (schemas.length === 1) {
            return schemas[0];
        }
        const node = epoxy_1.mergeTypeDefs(schemas);
        return graphql_1.buildASTSchema(node);
    }
    isSchemaText(obj) {
        return typeof obj === 'string';
    }
    isWrappedSchemaJson(obj) {
        const json = obj;
        return json.data !== undefined && json.data.__schema !== undefined;
    }
    isSchemaJson(obj) {
        const json = obj;
        return json !== undefined && json.__schema !== undefined;
    }
    isSchemaObject(obj) {
        return obj instanceof graphql_1.GraphQLSchema;
    }
    isSchemaAst(obj) {
        return obj.kind !== undefined;
    }
    isPromise(obj) {
        return !!obj && (typeof obj === 'object' || typeof obj === 'function') && typeof obj.then === 'function';
    }
    async resolveSchema(schema) {
        if (this.isSchemaObject(schema)) {
            return schema;
        }
        else if (this.isSchemaText(schema)) {
            const ast = graphql_1.parse(schema);
            return graphql_1.buildASTSchema(ast);
        }
        else if (this.isWrappedSchemaJson(schema)) {
            return graphql_1.buildClientSchema(schema.data);
        }
        else if (this.isSchemaJson(schema)) {
            return graphql_1.buildClientSchema(schema);
        }
        else if (this.isSchemaAst(schema)) {
            return schema;
        }
        else {
            throw new Error('Unexpected schema type provided!');
        }
    }
}
exports.SchemaFromExport = SchemaFromExport;
